src/GafferUSD/USDLayerWriter.cpp

```diff
@@ -68,6 +75,146 @@ using namespace GafferUSD;
 namespace
 {
 
+const PathMatcher g_emptyPathMatcher;
+
+struct Filters
+{
+	bool fullyPruned = true; // True only if all descendants present in `prune`.
+	PathMatcher prune;
+	PathMatcher deleteObject;
+	PathMatcher deleteAttributes;
+
+	// Merges in filters from sibling locations.
+	void add( const Filters &other )
+	{
+		fullyPruned = fullyPruned && other.fullyPruned;
+		prune.addPaths( other.prune );
+		deleteAttributes.addPaths( other.deleteAttributes );
+		deleteObject.addPaths( other.deleteObject );
+	}
+
+	// Merges in filters from child locations.
+	void addWithPrefix( const Filters &other, const ScenePlug::ScenePath &prefix )
+	{
+		fullyPruned = fullyPruned && other.fullyPruned;
+		prune.addPaths( other.prune, prefix );
+		deleteAttributes.addPaths( other.deleteAttributes, prefix );
+		deleteObject.addPaths( other.deleteObject, prefix );
+	}
+
+};
+
+// Recursively builds the PathMatchers held in a Filters struct, such that :
+//
+// - Objects with identical hashes will be included in `Filter::deleteObject`.
+// - Attributes with identical hashes will be included in `Filter::deleteAttributes`.
+// - Subtrees which are identical in all properties will be included in `Filter::prune`.
+Filters filtersWalk( const GafferScene::ScenePlug *baseScene, const GafferScene::ScenePlug *layerScene, const std::vector<float> &frames, const ScenePlug::ScenePath &path, const Gaffer::ThreadState &threadState, tbb::task_group_context &taskGroupContext )
+{
+	ScenePlug::PathScope pathScope( threadState, &path );
+
+	if( !layerScene->existsPlug()->getValue() )
+	{
+		return { false, g_emptyPathMatcher, g_emptyPathMatcher, g_emptyPathMatcher };
+	}
+
+	bool attributesMatch = true;
+	bool objectsMatch = true;
+	bool canPrune = true;
+
+	for( auto frame : frames )
+	{
+		pathScope.setFrame( frame );
+
+		if( attributesMatch )
+		{
+			attributesMatch = baseScene->attributesPlug()->hash() == layerScene->attributesPlug()->hash();
+			canPrune = canPrune && attributesMatch;
+		}
+
+		if( objectsMatch )
+		{
+			objectsMatch = baseScene->objectPlug()->hash() == layerScene->objectPlug()->hash();
+			canPrune = canPrune && objectsMatch;
```

Comment:
Just an aesthetics thing, but I think I would find this a bit clearer if the first two assignments to canPrune were replaced with `canPrune = canPrune && attributesMatch && objectsMatch` in the canPrune branch below.